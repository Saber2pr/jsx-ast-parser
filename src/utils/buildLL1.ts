import { buildLexer } from 'typescript-parsec'

export const buildLL1: (words: string[], kind: any) => typeof buildLexer =
  (words, kind) => rules => {
    const conflicts: any[][] = []
    for (const word of words) {
      const matched = []
      for (const rule of rules) {
        if (word.match(rule[1])) {
          matched.push(kind[rule[2]])
        }
      }
      if (matched.length > 1) {
        conflicts.push([word, ...matched])
      }
    }
    if (conflicts.length) {
      console.log(
        `[LL1 Conflicts] \n${conflicts
          .map(row => `${row[0]}: ${row.slice(1).join(',')}`)
          .join('\n')}`
      )
      process.exit(1)
    }
    return buildLexer(rules)
  }
