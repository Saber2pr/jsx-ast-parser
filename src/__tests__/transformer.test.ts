/*
 * @Author: saber2pr
 * @Date: 2021-09-12 12:06:14
 * @Last Modified by: saber2pr
 * @Last Modified time: 2021-10-07 11:00:34
 */
import { parser, transformer } from '../'
import { read } from './utils'

describe('Transformer', () => {
  it('Case 1', async () => {
    const code = await read('code.txt')
    expect(transformer.transform(parser.parse(code))).toMatchSnapshot()
  })

  const transform = (code: string) =>
    transformer.transform(parser.parse(code).body[0])

  it('CallChainExpr', () => {
    expect(transformer.isCallChain(transform('console.log(error)'))).toBe(true)
  })

  it('DefineVariableStatement', () => {
    expect(transformer.isDefineVariable(transform('const a = "a"'))).toBe(true)
  })

  it('VariableAssignExpr', () => {
    expect(transformer.isVariableAssign(transform('a = "a"'))).toBe(true)
  })

  it('IfStatement', () => {
    expect(
      transformer.isIf(
        transform(`
        if(a){
          console.log(a)
        } else if(b) {
          console.log(b)
        } else {
          console.log(c)
        }
        `)
      )
    ).toBe(true)
  })

  it('ReturnStatement', () => {
    expect(
      transformer.isReturn(
        transform(`
          return 233
        `)
      )
    ).toBe(true)
  })

  it('JsxExpr', () => {
    expect(transformer.isJsxElement(transform('<div>233</div>'))).toBe(true)
  })

  it('JsxSelfClosingExpr', () => {
    expect(transformer.isJsxElement(transform('<div className="233" />'))).toBe(
      true
    )
  })

  it('StringExpr', () => {
    expect(transform('"saber2pr"')).toBe('saber2pr')
  })

  it('NumberExpr', () => {
    expect(transform('233')).toBe(233)
  })

  it('ObjectExpr', () => {
    expect(transformer.isJsxObject(transform('{a:"a"}'))).toBe(true)
  })

  it('ArrayExpr', () => {
    expect(transform('["a", "b"]')).toEqual(['a', 'b'])
  })

  it('ArrowFunctionExpr', () => {
    expect(
      transformer.isArrowFunction(transform('(a) => { console.log(a) }'))
    ).toBe(true)
  })

  it('FunctionExpr', () => {
    expect(
      transformer.isFunction(transform('function test(a){ console.log(a) }'))
    ).toBe(true)
  })
})
