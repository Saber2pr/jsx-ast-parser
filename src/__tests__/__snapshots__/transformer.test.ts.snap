// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transformer Case 1 1`] = `
Array [
  Object {
    "children": Array [
      Object {
        "children": Array [],
        "props": Object {
          "list": Array [
            Object {
              "content": Object {
                "children": Array [
                  Object {
                    "nodeValue": "233",
                    "tagName": "text",
                  },
                ],
                "props": Object {
                  "color": "red",
                },
                "tagName": "View",
              },
              "logo": Object {
                "children": Array [],
                "props": Object {
                  "mode": "test",
                },
                "tagName": "Image",
              },
            },
            Object {
              "content": Object {
                "children": Array [],
                "props": Object {},
                "tagName": "View",
              },
            },
          ],
        },
        "tagName": "List",
      },
      Object {
        "children": Array [],
        "props": Object {},
        "tagName": "div",
      },
      Object {
        "children": Array [],
        "props": Object {
          "id": "qwq",
        },
        "tagName": "div",
      },
      Object {
        "children": Array [
          Object {
            "nodeValue": "aaa",
            "tagName": "text",
          },
        ],
        "props": Object {},
        "tagName": "span",
      },
      Object {
        "children": Array [
          Object {
            "nodeValue": "1234",
            "tagName": "text",
          },
        ],
        "props": Object {},
        "tagName": "span",
      },
      Object {
        "children": Array [
          Object {
            "nodeValue": "1234asd",
            "tagName": "text",
          },
        ],
        "props": Object {},
        "tagName": "span",
      },
      Object {
        "children": Array [
          Object {
            "nodeValue": "12aa",
            "tagName": "text",
          },
          Object {
            "children": Array [
              Object {
                "nodeValue": "aaa",
                "tagName": "text",
              },
            ],
            "props": Object {},
            "tagName": "span",
          },
          Object {
            "nodeValue": "aa234234aaa",
            "tagName": "text",
          },
        ],
        "props": Object {},
        "tagName": "span",
      },
    ],
    "props": Object {
      "arr": Array [
        1,
        2,
      ],
      "class2Name": "qwq123",
      "color": "red",
      "contentEditable": Object {
        "name": "true",
      },
      "id": "233ccc",
      "onClick": Object {
        "name": "onClick",
      },
      "onError": Object {
        "args": Array [
          Object {
            "name": "error",
          },
          Object {
            "name": "test",
          },
        ],
        "body": Object {
          "statements": Array [
            Object {
              "args": Array [
                Object {
                  "name": "error",
                },
              ],
              "caller": "console",
              "chain": Array [
                "log",
              ],
            },
            Object {
              "args": Array [],
              "caller": "console",
              "chain": Array [
                "log",
              ],
            },
            Object {
              "args": Array [
                Object {
                  "args": Array [
                    Object {
                      "name": "test",
                    },
                  ],
                  "caller": "console",
                  "chain": Array [
                    "log",
                  ],
                },
              ],
              "caller": "console",
              "chain": Array [
                "log",
              ],
            },
            Object {
              "assign": Object {
                "name": "a",
                "value": "a",
              },
              "type": "const",
            },
            Object {
              "assign": Object {
                "name": "b",
                "value": "b",
              },
              "type": "let",
            },
            Object {
              "assign": Object {
                "name": "c",
                "value": "c",
              },
              "type": "const",
            },
            Object {
              "assign": Object {
                "name": "d",
              },
              "type": "let",
            },
            Object {
              "name": "d",
              "value": "d",
            },
            Object {
              "assign": Object {
                "name": "a",
              },
              "type": "let",
            },
            Object {
              "name": "a",
              "value": "a",
            },
            Object {
              "args": Array [
                Object {
                  "name": "a",
                },
              ],
              "body": Object {
                "statements": Array [
                  Object {
                    "args": Array [
                      Object {
                        "name": "a",
                      },
                    ],
                    "caller": "console",
                    "chain": Array [
                      "log",
                    ],
                  },
                ],
              },
              "els": undefined,
            },
            Object {
              "args": Array [
                Object {
                  "name": "b",
                },
              ],
              "body": Object {
                "statements": Array [
                  Object {
                    "args": Array [
                      Object {
                        "name": "b",
                      },
                    ],
                    "caller": "console",
                    "chain": Array [
                      "log",
                    ],
                  },
                ],
              },
              "els": Object {
                "statements": Array [
                  Object {
                    "args": Array [
                      Object {
                        "name": "c",
                      },
                    ],
                    "caller": "console",
                    "chain": Array [
                      "log",
                    ],
                  },
                ],
              },
            },
            Object {
              "args": Array [
                Object {
                  "name": "b",
                },
              ],
              "body": Object {
                "statements": Array [
                  Object {
                    "args": Array [
                      Object {
                        "name": "b",
                      },
                    ],
                    "caller": "console",
                    "chain": Array [
                      "log",
                    ],
                  },
                ],
              },
              "els": Object {
                "args": Array [
                  Object {
                    "name": "c",
                  },
                ],
                "body": Object {
                  "statements": Array [
                    Object {
                      "args": Array [
                        Object {
                          "name": "c",
                        },
                      ],
                      "caller": "console",
                      "chain": Array [
                        "log",
                      ],
                    },
                  ],
                },
                "els": undefined,
              },
            },
            Object {
              "args": Array [
                Object {
                  "name": "b",
                },
              ],
              "body": Object {
                "statements": Array [
                  Object {
                    "args": Array [
                      Object {
                        "name": "b",
                      },
                    ],
                    "caller": "console",
                    "chain": Array [
                      "log",
                    ],
                  },
                ],
              },
              "els": Object {
                "args": Array [
                  Object {
                    "name": "c",
                  },
                ],
                "body": Object {
                  "statements": Array [
                    Object {
                      "args": Array [
                        Object {
                          "name": "c",
                        },
                      ],
                      "caller": "console",
                      "chain": Array [
                        "log",
                      ],
                    },
                  ],
                },
                "els": Object {
                  "statements": Array [
                    Object {
                      "args": Array [
                        Object {
                          "name": "d",
                        },
                      ],
                      "caller": "console",
                      "chain": Array [
                        "log",
                      ],
                    },
                  ],
                },
              },
            },
          ],
        },
      },
      "onSubmit": Object {
        "args": Array [],
        "body": Object {
          "statements": Array [
            Object {
              "args": Array [],
              "caller": "console",
              "chain": Array [
                "log",
              ],
            },
          ],
        },
        "name": "test",
      },
      "style": Object {
        "background": "blue",
        "child": Object {
          "children": Array [
            Object {
              "nodeValue": "233",
              "tagName": "text",
            },
          ],
          "props": Object {},
          "tagName": "span",
        },
        "color": "red",
        "test": Object {
          "color": "red",
        },
        "width": 100,
      },
      "width": 100,
    },
    "tagName": "div",
  },
]
`;
